# Coding Exercise
Below you'll find a prompt describing a function that you need to implement.

1. **Instead of changing existing code in place, create a new cell and rewrite the code.** This is important because we can't see the history of edited cells in the notebook.
2. **Before you write code, explain what you’re about to do.** For example, if the previous cell has an error that you want to fix, explain what caused the error and how you plan to fix it.
3. **Before you execute code that has outputs, think about what you hope to learn.** This notebook will prompt you write down your thought process. Do not update your explanation after seeing the output, unless you made a genuine mistake when guessing (e.g. typo).

This notebook will prompt you with questions. Please answer them in the markdown cell. Do not modify the questions. Your answers should be specific to the code that you are writing, so please avoid generic answers. For example, if asked, "What do you plan to do?", please avoid simply answering, "I will test the code." Instead explain how you will test the code and why (e.g. "I will now test the edge case of an empty list, which should return...").

## Task 1 Prompt

Have you heard of the fibonacci sequence? It is defined by $f_0 = 0, f_1 = 1$ and $f_n=f_{n-1}+ f_{n-2}$ for $n\geq 2$.

This question is about a similar sequence, called gibonacci sequence. It is defined by $g_0 = x, g_1 = y$ and $g_n=g_{n-1}- g_{n-2}$ for $n\geq 2$. Different possible starting values of the gibonacci sequence may lead to different gibonacci sequences.

### Task
You are given three arguments: `n, x, y`. Your task is to find $g_n$ using the above definition of a gibonacci sequence starting with $g_0=x, g_1=y$

Write a function `gibonacci(n, x, y)` that takes in 3 arguments.

### Inputs

It is guaranteed that $n,x,y$ are integers and $0 \leq n,x,y \leq 10^9$.

### Outputs

`gibonacci(n, x, y)` should return $g_n$ given that $g_0=x, g_1=y$.

### Examples

`gibonacci(0, 0, 1) = 0`

`gibonacci(1, 0, 1) = 1`

`gibonacci(2, 0, 1) = 1`

`gibonacci(3, 0, 1) = 0`

TASK STARTED AT 12:05 PM

# **Q: How will you begin?**

I will begin by defining the function and the parameters it will take in. I will then define the base cases and the recursive case and there after return the recursive case.

In [3]:
# base case is 
# g(0) = x
# g(1) = y
# recursive case is
# g(n) = g(n-1) + g(n-2)

def gibonacci(n, x, y):
    if n == 0:
        return x
    elif n == 1:
        return y
    else:
        return gibonacci(n-1, x, y) + gibonacci(n-2, x, y)

gibonacci(0, 0, 1)


0

In [ ]:
# executed in 0.3s

In [5]:
print(gibonacci(4, 0, 1))

3


In [ ]:
# executed in 0.1s

**Q: What do you want to learn from running this code?**

I learn that the gibonacci sequence is a sequence of numbers where each number is the sum of the previous two numbers

**Q: What is your plan now?**

My plan is to make the code more efficient by using memoization. I will do this by using a dictionary to store the values of the function. I will then check if the value is in the dictionary. If it is, I will return the value. If it is not, I will calculate the value and store it in the dictionary. I will then return the value.



In [6]:
def gibonacci(n, x, y, memo = {}):
    if n == 0:
        return x
    elif n == 1:
        return y
    else:
        if n in memo:
            return memo[n]
        else:
            memo[n] = gibonacci(n-1, x, y) + gibonacci(n-2, x, y)
            return memo[n]
#call function
gibonacci(10, 1, 2)

144

**Q: What do you expect the output to be (if any)?**

144

**Q: Explain any problems with the code (skip if not applicable).**


**Q: What do you want to learn from running this code?**

I want to learn how to use recursion to solve problems

**Q: What will you do next? Why?**

I will test the function to see if it works. I will do this by calling the function with different values of n, x, y.



In [4]:
print(gibonacci(0, 0, 1))
print(gibonacci(1, 0, 1))
print(gibonacci(2, 0, 1))
print(gibonacci(3, 0, 1))

0
1
1
2


In [ ]:
# executed in 0.5s

**Q: What are you trying to accomplish with this code?**

I am trying to run extra tests to ensure that the  function passes all tests.

**Q: Explain any problems with the code (skip if not applicable).**

The code works fine. It is a recursive function that returns the g(n) of the gibonacci sequence.



**Q: What are you thinking?**

I am thinking of how I can improve the code.

How I can make the code more efficient. 

How I can make the code more readable

ENDED AT 12:51 PM